Sprint 7 solution in file manager

masmetus opened the pull request.

masmetus committed:
Первая итерация ФЗ 6 спринта
Реализован кастомный связанный список и работа с ним. Так же метод удаления по id в интерфейсе

masmetus committed:
fix after CheckStyle

masmetus committed:
fix v2 after CheckStyle

masmetus committed:
fix v3 after CheckStyle

masmetus committed:
fix test after create new logic for saving tasks in history

masmetus committed:
fix after review:
1. remove history limit
2. fix methode after remove limit
3. moved getId to a variable

masmetus committed:
Реализовано:
1. Сохранение задач в файл. И класс обработчик ошибок
2. Доп задание - пользовательский сценарий
3. Юнит тесты

masmetus committed:
Фикс стиля

masmetus committed:
Фикс стиля v2

z8q commented on src/utils/Managers.java:
```diff
@@ -7,7 +7,7 @@
 
-    public static TaskManager getDefault(){
+    public static TaskManager getDefault() {
         return new InMemoryTaskManager();
```
тут можно теперь вернуть FileBackedTaskManager в качестве дефолта

z8q commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,249 @@
+            }
+
+        } catch (FileNotFoundException e) {
+            return manager;
```
возможно будет правильно бросить исключение

потому что без файла мы не сможем полноценно использовать менеджер

если захотим что-то добавить/изменить, нам нужен будет файл для записи, а его нет

z8q commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,249 @@
+                if (task instanceof Epic) {
+                    manager.createEpic((Epic) task);
+                    epicSubtasks.put(task.getId(), new ArrayList<>());
+                } else if (task instanceof Subtask) {
```
в новых версиях java можно писать таким образом
if (task instanceof Subtask subtask)
и тогда 114-ю строку можно будет не писать

z8q commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,249 @@
+    }
+
+
+    private Task fromCsvString(String csvLine) {
```
методы fromCsvString  и toCsvString можно вынести в отдельный класс
например CsvTaskMapper
и сделать статическими, чтобы к ним можно было обращаться таким образом
CsvTaskMapper.fromCsvString(...)
CsvTaskMapper.toCsvString(...)

masmetus committed:
Исправления по итогам реквеста:
1. Перенёс часть логики в отельный класс CsvTaskMapper
2. Оптимизация кода, ввиду использования новой версии java
3. Удалил обработку отсутствия файла
4. Подправил getDefault в классе Managers
